opensearch_acls.go

```diff
@@ -0,0 +1,129 @@
+	// https://api.aiven.io/v1/project/<project>/service/<service_name>/opensearch/acl
+	OpenSearchACLRequest struct {
+		OpenSearchACLConfig OpenSearchACLConfig `json:"opensearch_acl_config"`
+	}
+
+	// OpenSearchACLResponse Aiven API response
+	// https://api.aiven.io/v1/project/<project>/service/<service_name>/opensearch/acl
+	OpenSearchACLResponse struct {
+		APIResponse
+		OpenSearchACLConfig OpenSearchACLConfig `json:"opensearch_acl_config"`
+	}
+
+	// OpenSearchACLConfig represents a configuration for OpenSearch ACLs
+	OpenSearchACLConfig struct {
+		ACLs        []OpenSearchACL `json:"acls"`
+		Enabled     bool            `json:"enabled"`
+		ExtendedAcl bool            `json:"extendedAcl"`
+	}
+
+	// OpenSearchACL represents a OpenSearch ACLs entry
+	OpenSearchACL struct {
+		Rules    []OpenSearchACLRule `json:"rules"`
+		Username string              `json:"username"`
+	}
+
+	// OpenSearchACLRule represents a OpenSearch ACLs Rule entry
+	OpenSearchACLRule struct {
+		Index      string `json:"index"`
+		Permission string `json:"permission"`
+	}
+)
+
+// Update updates OpenSearch ACL config
+func (h *OpenSearchACLsHandler) Update(ctx context.Context, project, service string, req OpenSearchACLRequest) (*OpenSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "opensearch", "acl")
+	bts, err := h.client.doPutRequest(ctx, path, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var r OpenSearchACLResponse
+	errR := checkAPIResponse(bts, &r)
+
+	return &r, errR
+}
+
+// Get gets all existing OpenSearch ACLs config
+func (h *OpenSearchACLsHandler) Get(ctx context.Context, project, service string) (*OpenSearchACLResponse, error) {
+	path := buildPath("project", project, "service", service, "opensearch", "acl")
+	bts, err := h.client.doGetRequest(ctx, path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var r OpenSearchACLResponse
+	errR := checkAPIResponse(bts, &r)
+
+	return &r, errR
+}
+
+// Delete subtracts ACL from already existing OpenSearch ACLs config
+func (conf *OpenSearchACLConfig) Delete(ctx context.Context, acl OpenSearchACL) *OpenSearchACLConfig {
+	for p, existingAcl := range conf.ACLs { // subtract ALC from existing ACLs config entry that supposed to be deleted
+		if acl.Username == existingAcl.Username {
+			for i := range existingAcl.Rules {
+				// remove ACL from existing ACLs list
+				for _, rule := range acl.Rules {
+					if existingAcl.Rules[i].Permission == rule.Permission && existingAcl.Rules[i].Index == rule.Index {
+						conf.ACLs[p].Rules = append(conf.ACLs[p].Rules[:i], conf.ACLs[p].Rules[i+1:]...)
+					}
+				}
+
+				// delete ACL item from ACLs list is there are not rules attached to it
+				if len(conf.ACLs[p].Rules) == 0 {
+					conf.ACLs = append(conf.ACLs[:p], conf.ACLs[p+1:]...)
+				}
+			}
+		}
+	}
+
+	return conf
+}
+
+// Add appends new ACL to already existing OpenSearch ACLs config
+func (conf *OpenSearchACLConfig) Add(acl OpenSearchACL) *OpenSearchACLConfig {
```

Comment:
this is also quite complex, how about that:

```go
// Add appends new ACL to the existing OpenSearch ACLs config.
func (conf *OpenSearchACLConfig) Add(acl OpenSearchACL) *OpenSearchACLConfig {
	var userIndex int
	userExists := false

	// Iterate over the existing ACLs to identify duplicates and determine user existence.
	for p, existingAcl := range conf.ACLs {
		if acl.Username == existingAcl.Username {
			userExists = true
			userIndex = p

			// Filter out any rules in the ACL to add that already exist for the user.
			remainingRules := []Rule{}
			for _, rule := range acl.Rules {
				exists := false
				for _, existingRule := range existingAcl.Rules {
					if rule.Permission == existingRule.Permission && rule.Index == existingRule.Index {
						exists = true
						break
					}
				}
				if !exists {
					remainingRules = append(remainingRules, rule)
				}
			}
			acl.Rules = remainingRules
		}
	}

	// If no rules remain for the user, return the existing configuration.
	if len(acl.Rules) == 0 {
		return conf
	}

	// Add the new or updated ACL to the config.
	if userExists {
		conf.ACLs[userIndex].Rules = append(conf.ACLs[userIndex].Rules, acl.Rules...)
	} else {
		conf.ACLs = append(conf.ACLs, acl)
	}

	return conf
}
```

Reply:
Done